app/src/main/java/org/mozilla/fenix/gleanplumb/state/MessagingMiddleware.kt

```diff
@@ -0,0 +1,85 @@
+
+package org.mozilla.fenix.gleanplumb.state
+
+import mozilla.components.lib.state.Middleware
+import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.components.appstate.AppAction
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.Initialize
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageClicked
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageDismissed
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.UpdateMessages
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.MessageDisplayed
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.Evaluate
+import org.mozilla.fenix.components.appstate.AppAction.MessagingAction.UpdateMessageToShow
+import org.mozilla.fenix.components.appstate.AppState
+import org.mozilla.fenix.gleanplumb.Message
+import org.mozilla.fenix.gleanplumb.MessagingStorage
+
+class MessagingMiddleware(private val storage: MessagingStorage) : Middleware<AppState, AppAction> {
+
+    override fun invoke(
+        context: MiddlewareContext<AppState, AppAction>,
+        next: (AppAction) -> Unit,
+        action: AppAction
+    ) {
+        when (action) {
+            is Evaluate -> {
+                val nextMessage = storage.getNextMessage()
+                context.dispatch(UpdateMessageToShow(nextMessage))
+            }
+            is Initialize -> {
+                storage.init()
+            }
+            is MessageClicked -> {
+                // Update Nimbus storage.
+                val updatedMetadata = action.message.metadata.copy(pressed = true)
+                storage.updateMetadata(updatedMetadata)
+
+                // Update app state.
+                val newMessages = removeMessage(context, action.message)
+                context.dispatch(UpdateMessages(newMessages))
+            }
+            is MessageDismissed -> {
+                val newMessages = removeMessage(context, action.message)
+                val updatedMetadata = action.message.metadata.copy(dismissed = true)
+
+                storage.updateMetadata(updatedMetadata)
+                context.dispatch(UpdateMessages(newMessages))
+            }
+            is MessageDisplayed -> {
+                val oldMessage = action.message
+                val newMetadata = oldMessage.metadata.copy(
+                    displayCount = oldMessage.metadata.displayCount + 1
+                )
+                val newMessage = oldMessage.copy(
+                    metadata = newMetadata
+                )
+                val newMessages = if (newMetadata.displayCount < oldMessage.data.maxDisplayCount) {
+                    updateMessage(context, oldMessage, newMessage)
+                } else {
+                    removeMessage(context, oldMessage)
+                }
+                context.dispatch(UpdateMessages(newMessages))
```

Comment:
Yes, that's right! If you don't need this to happen synchronously `store.dispatch()` will do. Dispatching on the current context is useful in cases where the middleware wants to trigger additional state changes before continuing. Also, important to point out that `context.dispatch` needs to happen on the current thread and will otherwise crash to prevent incorrect usage.